File: backend/src/controllers/profile.controller.ts
import { Request, Response } from 'express'
import { supabase } from '../config/supabase'

// ──────────────────────────────────────────────────────────
// GET /api/profile/me
// Retorna o perfil do usuário que está logado no momento.
// ──────────────────────────────────────────────────────────
export const getMyProfile = async (req: Request, res: Response) => {
  // req.user vem do middleware de autenticação (auth.middleware.ts)
  const userId = req.user!.id

  const { data, error } = await supabase
    .from('profiles')
    .select('*')       // pega todos os campos do perfil
    .eq('id', userId)  // filtra pelo id do usuário logado
    .single()          // retorna um único objeto, não um array

  if (error) {
    return res.status(404).json({ error: 'Perfil não encontrado' })
  }

  return res.status(200).json({ profile: data })
}

// ──────────────────────────────────────────────────────────
// PUT /api/profile/me
// Atualiza os dados do perfil do usuário logado.
// Campos permitidos: name, phone, avatar_url, bio
// O campo "role" NÃO pode ser alterado pelo próprio usuário.
// ──────────────────────────────────────────────────────────
export const updateMyProfile = async (req: Request, res: Response) => {
  const userId = req.user!.id

  // Extrai apenas os campos que o usuário tem permissão de alterar
  const { name, phone, avatar_url, bio } = req.body

  // Monta o objeto de atualização apenas com os campos enviados
  const updates: Record<string, unknown> = {}
  if (name       !== undefined) updates.name       = name
  if (phone      !== undefined) updates.phone      = phone
  if (avatar_url !== undefined) updates.avatar_url = avatar_url
  if (bio        !== undefined) updates.bio        = bio

  if (Object.keys(updates).length === 0) {
    return res.status(400).json({ error: 'Nenhum campo para atualizar foi enviado' })
  }

  const { data, error } = await supabase
    .from('profiles')
    .update(updates)
    .eq('id', userId)  // garante que só atualiza o próprio perfil
    .select()
    .single()

  if (error) {
    return res.status(400).json({ error: error.message })
  }

  return res.status(200).json({ profile: data })
}

// ──────────────────────────────────────────────────────────
// GET /api/profile/all  [somente admin]
// Retorna a lista de todos os usuários cadastrados.
// ──────────────────────────────────────────────────────────
export const getAllProfiles = async (_req: Request, res: Response) => {
  const { data, error } = await supabase
    .from('profiles')
    .select('id, name, email, role, avatar_url, created_at')
    .order('created_at', { ascending: false }) // mais recentes primeiro

  if (error) {
    return res.status(400).json({ error: error.message })
  }

  return res.status(200).json({ profiles: data })
}

// ──────────────────────────────────────────────────────────
// PUT /api/profile/:id/role  [somente admin]
// Altera o papel (role) de qualquer usuário.
// Exemplo: promover para 'pro' ou 'admin'.
// ──────────────────────────────────────────────────────────
export const updateUserRole = async (req: Request, res: Response) => {
  const { id } = req.params  // id do usuário alvo (da URL)
  const { role } = req.body  // novo papel enviado no corpo da requisição

  // Valida se o papel enviado é um dos valores permitidos
  const rolesPermitidos = ['user', 'pro', 'admin']
  if (!rolesPermitidos.includes(role)) {
    return res.status(400).json({
      error: `Role inválido. Use: ${rolesPermitidos.join(', ')}`
    })
  }

  const { data, error } = await supabase
    .from('profiles')
    .update({ role })
    .eq('id', id)
    .select()
    .single()

  if (error) {
    return res.status(400).json({ error: error.message })
  }

  return res.status(200).json({ profile: data })
}

File: backend-go/internal/profile/profile_controller.go
package profile

import (
	"fmt"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"beeplanerofc/backend-go/internal/dbutil"
	"beeplanerofc/backend-go/internal/middleware"
)

type Controller struct {
	db *pgxpool.Pool
}

func NewController(db *pgxpool.Pool) *Controller {
	return &Controller{db: db}
}

// GET /api/profile/me — retorna o perfil do usuário logado
func (ctrl *Controller) GetMyProfile(c *gin.Context) {
	user := middleware.CurrentUser(c)

	rows, err := ctrl.db.Query(c.Request.Context(), "SELECT * FROM profiles WHERE id = $1", user.ID)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Perfil não encontrado"})
		return
	}

	profileData, err := pgx.CollectOneRow(rows, dbutil.RowToMap)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Perfil não encontrado"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"profile": profileData})
}

type updateProfileRequest struct {
	Name      *string `json:"name"`
	Phone     *string `json:"phone"`
	AvatarURL *string `json:"avatar_url"`
	Bio       *string `json:"bio"`
	Birthdate *string `json:"birthdate"`
}

// PUT /api/profile/me — atualiza apenas os campos enviados do perfil logado.
// O campo "role" não pode ser alterado pelo próprio usuário.
func (ctrl *Controller) UpdateMyProfile(c *gin.Context) {
	user := middleware.CurrentUser(c)

	var req updateProfileRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Dados inválidos"})
		return
	}

	setClauses := []string{}
	args := []any{}
	addField := func(column string, value *string) {
		if value == nil {
			return
		}
		args = append(args, *value)
		setClauses = append(setClauses, fmt.Sprintf("%s = $%d", column, len(args)))
	}

	addField("name", req.Name)
	addField("phone", req.Phone)
	addField("avatar_url", req.AvatarURL)
	addField("bio", req.Bio)
	addField("birthdate", req.Birthdate)

	if len(setClauses) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Nenhum campo para atualizar foi enviado"})
		return
	}

	args = append(args, user.ID)
	query := fmt.Sprintf(
		"UPDATE profiles SET %s WHERE id = $%d RETURNING *",
		strings.Join(setClauses, ", "), len(args),
	)

	rows, err := ctrl.db.Query(c.Request.Context(), query, args...)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	profileData, err := pgx.CollectOneRow(rows, dbutil.RowToMap)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"profile": profileData})
}

// GET /api/profile/all [somente admin] — lista todos os usuários
func (ctrl *Controller) GetAllProfiles(c *gin.Context) {
	rows, err := ctrl.db.Query(c.Request.Context(),
		"SELECT id, name, email, role, avatar_url, created_at FROM profiles ORDER BY created_at DESC",
	)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	profiles, err := pgx.CollectRows(rows, dbutil.RowToMap)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"profiles": profiles})
}

type updateRoleRequest struct {
	Role string `json:"role"`
}

var rolesPermitidos = map[string]bool{"user": true, "pro": true, "admin": true}

// PUT /api/profile/:id/role [somente admin] — altera o papel de qualquer usuário
func (ctrl *Controller) UpdateUserRole(c *gin.Context) {
	id := c.Param("id")

	var req updateRoleRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Dados inválidos"})
		return
	}

	if !rolesPermitidos[req.Role] {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Role inválido. Use: user, pro, admin"})
		return
	}

	rows, err := ctrl.db.Query(c.Request.Context(),
		"UPDATE profiles SET role = $1 WHERE id = $2 RETURNING *", req.Role, id,
	)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	profileData, err := pgx.CollectOneRow(rows, dbutil.RowToMap)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"profile": profileData})
}
